bpf/bpf_overlay.c

```diff
@@ -194,11 +195,21 @@ static __always_inline int handle_ipv6(struct __ctx_buff *ctx,
 	}
 #endif /* ENABLE_EGRESS_GATEWAY_COMMON */
 
+#if defined(ENABLE_DSR) && (DSR_ENCAP_MODE == DSR_ENCAP_GENEVE)
+	/* Pass packets which will be returned using Geneve DSR
+	 * to host-stack for conntrack entry insertion.
+	 * This is needed because the packet will be masqueraded
+	 * by iptables if the conntrack entry isn't exist.
+	 */
+	if (!is_defined(ENABLE_HOST_ROUTING) && is_dsr)
+		is_delivered_via_stack = true;
+#endif
+
```

Comment:
I had another look at how this scenario is [handled](https://github.com/cilium/cilium/blob/e87e9e21d3c243c823e8852cecc56f7efc92e8b6/bpf/bpf_host.c#L724) when DSR traffic arrives in `from-netdev` (with `dispatch-mode=opt`). And there we exit *before* looking up the local endpoint, or calling into `ipv*_local_delivery()`. Which means the equivalent change here would be as seen below.

I think that's preferable - it reduces the amount of code locations we need to touch (no additional parameter needed for `ipv*_local_delivery()`), and ensures we have symmetric behavior with the other DSR mode.


```suggestion
#if defined(ENABLE_DSR) && (DSR_ENCAP_MODE == DSR_ENCAP_GENEVE)
	/* Pass packets which will be returned using Geneve DSR
	 * to host-stack for conntrack entry insertion.
	 * This is needed because the packet will be masqueraded
	 * by iptables if the conntrack entry doesn't exist.
	 */
	if (!is_defined(ENABLE_HOST_ROUTING) && is_dsr)
		return CTX_ACT_OK;
#endif
```

Reply:
I've fixed and confirmed that BPF unit tests and Geneve-DSR test workload in local-kind-cluster work well.